### ISSUE TITLE
Memory leak

### ISSUE THREAD

**Issue by caiocesarrm** [pymupdf/PyMuPDF#881]
Hi, really appreciate this excellent work!

I have an internal service running pymupdf to extract text and highlight pdf, i've been using since january 2020 always updating the versions. Sometimes i've noticed that the lib may have a memory leak, changing to a new version solved partially but later or sooner this possible memory leak goes back. The first time the memoryleak was in extract_text, but i believe that after i updated to a new version the memory leak has gone.

Now i have a memory leak. 

The workaround that i'm using is restarting the worker after 10 jobs. 

I used the python standard memory profilling lib (tracemallock).

Here is the first job
![first_run](https://user-images.githubusercontent.com/41218670/106854709-89edec80-669a-11eb-94dc-f8b4e58887f5.png)
Here is after 6 jobs
![mid](https://user-images.githubusercontent.com/41218670/106854714-8d817380-669a-11eb-88a7-c5a70b08f041.png)
After the final job
![final](https://user-images.githubusercontent.com/41218670/106854727-9114fa80-669a-11eb-8188-417e48ddc538.png)

as you can see, multiple lines add constant memory allocation from each run, like the line 6013 from fitz.py, adding 36.2 after each job. Each job is a task that runs independently, with the Celery Library. The line 6013 in fitz.py corresponds to insertImage, wich i'm using on job, i'm also using insertPage, insertTextbox, Document.select, getPixmap, document.Write, movePage and insertPDF.

I really don't know if i'm doing something wrong or if the problem is with pymupdf.

PyMuPDF 1.18.6.
Built for Python 3.6 on linux (64-bit).

**Comment by JorjMcKie** [comment]
Thanks for reporting this! It is really important to track down those possible leaks.
On the other hand, it is improbable that there are leaks all over the place. I know there is a minor one left over after several rounds of fixing in insertPDF - which is inevitable unfortunately.

I would like to start with insert image:
Can you tell me how you use it: pixmap, stream, from file?

**Comment by JorjMcKie** [comment]
Investigated insert_image:
This indeed has a memory leak for the file and stream insertion cases only. You should not see any leakages when inserting pixmaps. After changes, the following script
[gc-insertimage-stream.zip](https://github.com/pymupdf/PyMuPDF/files/5926355/gc-insertimage-stream.zip)

... converts every page of the Adobe manual (1'310 pages) to a memory image and inserts that in another PDF page. All that repeated 5 times. Gives this result:
```
PyMuPDF 1.18.8: Python bindings for the MuPDF 1.18.0 library.
Version date: 2021-02-03 19:56:11.
Built for Python 3.9 on win32 (64-bit).

File 'adobe.pdf'
Memory deltas (MB) for Page.insert_image from stream
0 stop delta 38.7
1 stop delta 4.4
2 stop delta -0.1
3 stop delta 0.9
4 stop delta 0.3
==================================================
Total pages processed: 6550
Duration: 175 sec
```
As you see, after the initial setup, practically no additional memory is leaked.
You may also notice that there is a function in `TOOLS` which allows resetting the internal cache held by MuPDF for various purposes. May also help from time to time.

**Comment by JorjMcKie** [comment]
Fixed in v1.18.8

**Comment by caiocesarrm** [comment]
It's no demagogy, you are a real legend, very thanks!
